pkg/volume/csi/csi_mounter.go

```diff
@@ -134,68 +150,119 @@ func (c *csiMountMgr) SetUpAt(dir string, fsGroup *int64) error {
+			volumeHandle = *volSrc.VolumeHandle
+		} else {
+			klog.Error(log("mounter.SetUpAt failed as VolumeHandle is nil", driverName))
+		}
+
+		if volSrc.FSType != nil {
+			fsType = *volSrc.FSType
+		}
+
+		volAttribs = volSrc.VolumeAttributes
+		klog.V(4).Info(log("mounter.SetUpAt getting PublishVolumeInfo"))
+		volumeInfo, err = c.plugin.getPublishVolumeInfo(c.k8s, volumeHandle, driverName, nodeName)
 		if err != nil {
+			klog.V(4).Info(log("mounter.SetUpAt failed to get VolumeInfo: %v", err))
 			return err
 		}
-	}
-
-	attribs := csiSource.VolumeAttributes
 
-	nodePublishSecrets := map[string]string{}
-	if csiSource.NodePublishSecretRef != nil {
-		nodePublishSecrets, err = getCredentialsFromSecret(c.k8s, csiSource.NodePublishSecretRef)
+		if volSrc.NodePublishSecretRef != nil {
+			secretName := volSrc.NodePublishSecretRef.Name
+			ns := c.pod.Namespace
+			nodePublishSecrets, err = getCredentialsFromSecret(c.k8s, &api.SecretReference{Name: secretName, Namespace: ns})
+			if err != nil {
+				return fmt.Errorf("fetching NodePublishSecretRef %s/%s failed: %v", ns, secretName, err)
+			}
+		}
+		mountOptions = []string{}
+
+		klog.V(4).Info(log("mounter.SetUpAt inline from CSIVolumeSource OK [driver:%d, volumeHandle:%s]", driverName, volumeHandle))
+	} else if pvSrc != nil {
+		// if source is a CSIPersistentVolumeSource (PV), prepare needed params (from PV)
+		klog.V(4).Info(log("mounter.SetUpAt from CSIPersistentVolumeSource"))
+		fsType = pvSrc.FSType
+		volAttribs = pvSrc.VolumeAttributes
+
+		// search for attachment by VolumeAttachment.Spec.Source.PersistentVolumeName
+		klog.V(4).Info(log("mounter.SetUpAt getting PublishVolumeInfo"))
+		c.volumeInfo, err = c.plugin.getPublishVolumeInfo(c.k8s, volumeHandle, driverName, nodeName)
 		if err != nil {
-			return fmt.Errorf("fetching NodePublishSecretRef %s/%s failed: %v",
-				csiSource.NodePublishSecretRef.Namespace, csiSource.NodePublishSecretRef.Name, err)
+			return err
 		}
-	}
+		volumeInfo = c.volumeInfo
 
-	// create target_dir before call to NodePublish
-	if err := os.MkdirAll(dir, 0750); err != nil {
-		klog.Error(log("mouter.SetUpAt failed to create dir %#v:  %v", dir, err))
-		return err
-	}
-	klog.V(4).Info(log("created target path successfully [%s]", dir))
+		// create target_dir before call to NodePublish
+		if err := os.MkdirAll(dir, 0750); err != nil {
+			klog.Error(log("mouter.SetUpAt failed to create dir %#v:  %v", dir, err))
+			return err
+		}
+		klog.V(4).Info(log("created target path successfully [%s]", dir))
+		if pvSrc.NodePublishSecretRef != nil {
+			nodePublishSecrets, err = getCredentialsFromSecret(c.k8s, pvSrc.NodePublishSecretRef)
+			if err != nil {
+				return fmt.Errorf("fetching NodePublishSecretRef %s/%s failed: %v",
+					pvSrc.NodePublishSecretRef.Namespace, pvSrc.NodePublishSecretRef.Name, err)
+			}
+		}
+
+		//TODO (vladimirvivien) implement better AccessModes mapping between k8s and CSI
+		if c.spec.PersistentVolume.Spec.AccessModes != nil {
```

Comment:
I think every time we only set a variable if its not nil I would like either an `else` statement for what it should do if it is nil, or a comment specifically stating that a nil variable (`accessMode` in this case) is ok. This seems to be a pattern throughout and its been very difficult to follow which variables are required, which are optional, which are missing error cases, etc.